feat: cancel query

CarbonChauvinist opened the pull request:
Allows user to cancel query if currently in Executing state.

Includes tests to ensure expected responses from server and that query manager remains in "Connected" state after cancellation.

<details>
<summary>example of VScode cancelation for reference:</summary>

    === CLIENT -> SERVER ===
    Content-Length: 92

    {"jsonrpc":"2.0","id":8,"method":"query/cancel","params":{"ownerUri":"untitled:Untitled-3"}}

    === SERVER -> CLIENT ===
    Content-Length: 53

    {"jsonrpc":"2.0","id":"8","result":{"messages":null}}

    === SERVER -> CLIENT ===
    Content-Length: 208

    {"jsonrpc":"2.0","method":"query/message","params":{"ownerUri":"untitled:Untitled-3","message":{"batchId":0,"isError":false,"time":"2025-06-14T18:12:33.4952694-04:00","message":"Query was canceled by user"}}}

    === SERVER -> CLIENT ===
    Content-Length: 430

    {"jsonrpc":"2.0","method":"query/batchComplete","params":{"batchSummary":{"executionElapsed":"00:00:05.3392796","executionEnd":"2025-06-14T18:12:33.5014254-04:00","executionStart":"2025-06-14T18:12:28.1621458-04:00","hasError":true,"id":0,"selection":{"endColumn":25,"endLine":2,"startColumn":0,"startLine":0},"resultSetSummaries":[],"specialAction":{"none":true,"expectYukonXMLShowPlan":false}},"ownerUri":"untitled:Untitled-3"}}

    === SERVER -> CLIENT ===
    Content-Length: 455

    {"jsonrpc":"2.0","method":"query/complete","params":{"ownerUri":"untitled:Untitled-3","batchSummaries":[{"executionElapsed":"00:00:05.3392796","executionEnd":"2025-06-14T18:12:33.5014254-04:00","executionStart":"2025-06-14T18:12:28.1621458-04:00","hasError":true,"id":0,"selection":{"endColumn":25,"endLine":2,"startColumn":0,"startLine":0},"resultSetSummaries":[],"specialAction":{"none":true,"expectYukonXMLShowPlan":false}}],"serverConnectionId":"74"}}
</details>

Fixes #64

CarbonChauvinist committed:
feat: cancel query

Allows user to cancel query if currently in Executing state.

Includes tests to ensure expected responses from server and that query
manager remains in "Connected" state after cancelation.

Fixes #64

CarbonChauvinist commented:
@Kurren123 Ok here's the first attempt at this. Do appreciate your patience and grace working through what I'm sure will be some back and forth. I've marked it as a draft for now in anticipation of the back and forth to clean this up.

My main concerns are the the tests - not sure how to actually run the tests - I think I've structured it correctly but am concerned about the asserts checking on the confirmation messages (I've included what I think is a server/client back and forth that I captured in VSCode for reference in the PR message above).

My other concern is how to ensure that the lualine/statusline integrations are working as expected. I think there's something else I'll need to do to show canceling while waiting for the cancellation (some times cancellations can take a while especially if the query is hung already).

TY in advance.

Kurren123 reviewed (changes requested):
Thank you for spending your time on this, much appreciated

CarbonChauvinist commented on lua/mssql/query_manager.lua:
```diff
@@ -119,6 +120,32 @@ return {
+
+				if err then
+					error("Could not cancel query: " .. vim.inspect(err), 0)
+				elseif not (result or result.batchSummaries) then
```
I'm not sure I know the answer to this, in my admittedly simple test earlier in VSC there was a result response but it appeared to be blank and had a null message? (see below). 

```
=== SERVER -> CLIENT ===
Content-Length: 53

{"jsonrpc":"2.0","id":"8","result":{"messages":null}}
```

This might be just a "response" from the initial query that was cancelled, and since it was cancelled there's actually no result to return but it still needs to respond to the original query request from the client? I have no idea honestly and that's just a wild guess.

What do you mean by enough to check if err is nil? Do you mean literally just something like this? 

```
if err then
    state.set_state(states.Executing)
    error("Error canceling query: " .. err.message, 0)
end
```

CarbonChauvinist committed:
fix(review): prevent duplicate listeners on query cancellation

The `execute` and `cancel` methods were both waiting for the same
`query/complete` notification, creating a potential conflict.

The `execute` method now checks if a cancellation is in progress. If so,
it returns early to let the `cancel` method handle the completion event.

CarbonChauvinist committed:
fix(review): restrict cancel action to Executing state

The 'cancel query' command and keymap are now only available while a
query is in the `Executing` state.

To enable this, the logic for generating available actions was
refactored from multiple return blocks into a single, modular
progressive-build styled pattern.

CarbonChauvinist committed:
test(review): simplify cancel query E2E test to verify state

CarbonChauvinist commented:
@Kurren123 Okay, I've pushed some commits that tries to address all of your valuable feedback. They should resolve all the conversations? It did end up with some changes to the underlying way the keymaps and usercommands are constructed which I'm hopeful was not a step too far. 

Ready for another look when you have the time, thanks!

CarbonChauvinist committed:
test(review): syntax error in cancel query test

CarbonChauvinist committed:
fix(review): race condition

Both `execute_async` and `cancel_async` are waiting for the same
'query/complete' notification.

- limit `cancel_async` to only setting state to 'Cancelling' and sending
  the 'query/cancel' request. It should not wait for a response
- original `execute_async` call, already waiting for 'query/complete',
  should handle the notification, check if state is 'Cancelling', and
  then finalize state change back to 'Connected'

CarbonChauvinist committed:
test(review): fix case of asserted value

States returned are all lowercase

Kurren123 reviewed (changes requested):
Great work, I think we're almost there. Only minor amendments now.

CarbonChauvinist committed:
fix(review): refine state transition after query completion

Refines how the query manager handles state transitions after query
execution, including potential cancellations, and avoids printing stack
trace for potential error messages.

CarbonChauvinist committed:
test(review): improve query manager access in cancel test

Streamlines accessing the query manager from buffer variables in cancel
query spec by removing an unnecessary call to get the current buffer.

CarbonChauvinist committed:
refactor(review): revert keymap and command DRY attempt for clarity

Reverts the previous attempt to consolidate keymap and user command
definitions, opting for explicit returns for each state. This aligns
with the preference for clearer, more direct readability and easier
debugging.

CarbonChauvinist committed:
docs: add CancelQuery keymap to documentation

Updates README.md to include 'CancelQuery' keymap in the list of
available keybindings.

CarbonChauvinist commented:
@Kurren123 Thanks for the heads up about what sounds to be an incredible PR, can't wait! 

I'm open to whatever is best from your perspective. I've just pushed the changes requested from your last review. I'm happy to wait until after you merge your branch and handle the conflicts that come up since my change is much smaller in scope.

(btw have you thought about setting up ability for github sponsors on this repo? With cross-db completion, this is literally the only tool that's allowed me to even think about doing my mssql day job in nvim and I'm sure others as well may be interested in helping support your work on this.)

Kurren123 committed:
Can only disconnect if state is connected

Kurren123 reviewed (approved):
Amazing, thanks for all your work!
